spectrum-offchain-cardano/src/data/dao_request.rs

```diff
@@ -0,0 +1,381 @@
+    pub dao_action: DaoAction,
+    pub pool_nft: PoolId,
+    pub pool_fee: Ratio<u64>,
+    pub treasury_fee: Ratio<u64>,
+    pub admin_address: ScriptHash,
+    pub pool_stake_script_hash: Option<ScriptHash>,
+    pub treasury_address: ScriptHash,
+    pub treasury_x_withdraw: TaggedAmount<Rx>,
+    pub treasury_y_withdraw: TaggedAmount<Ry>,
+    pub requestor_pkh: Ed25519KeyHash,
+    pub signatures: Vec<Ed25519Signature>,
+    pub additional_bytes: Vec<Vec<u8>>,
+    pub fee: Lovelace,
+    pub lovelace: Lovelace,
+}
+
+pub type OnChainDAOActionRequest = ClassicalOrder<OnChainOrderId, DAOV1Request>;
+
+impl Into<CFMMPoolAction> for OnChainDAOActionRequest {
+    fn into(self) -> CFMMPoolAction {
+        DAOAction
+    }
+}
+
+impl<Ctx> RequiresValidator<Ctx> for OnChainDAOActionRequest
+where
+    Ctx: Has<DeployedValidator<{ RoyaltyPoolDAOV1Request as u8 }>>,
+{
+    fn get_validator(&self, ctx: &Ctx) -> DeployedValidatorErased {
+        ctx.get().erased()
+    }
+}
+
+impl<Ctx> TryFromLedger<TransactionOutput, Ctx> for OnChainDAOActionRequest
+where
+    Ctx: Has<OutputRef> + Has<DeployedScriptInfo<{ RoyaltyPoolDAOV1Request as u8 }>>,
+{
+    fn try_from_ledger(repr: &TransactionOutput, ctx: &Ctx) -> Option<Self> {
+        if test_address(repr.address(), ctx) {
+            let pd = repr.datum().clone()?.into_pd()?;
+            let conf = DAORequestConfig::try_from_pd(pd)?;
+            let init_ada_value = repr.value().coin;
+            let royalty_withdraw = DAOV1Request {
+                dao_action: conf.dao_action,
+                pool_nft: PoolId::try_from(conf.pool_nft).ok()?,
+                pool_fee: Ratio::new_raw(conf.pool_fee, FEE_DEN),
+                treasury_fee: Ratio::new_raw(conf.treasury_fee, FEE_DEN),
+                admin_address: conf.admin_address,
+                pool_stake_script_hash: conf.pool_stake_script_hash,
+                treasury_address: conf.treasury_address,
+                treasury_x_withdraw: TaggedAmount::new(conf.treasury_x_withdraw),
+                treasury_y_withdraw: TaggedAmount::new(conf.treasury_y_withdraw),
+                requestor_pkh: conf.requestor_pkh,
+                signatures: conf.signatures,
+                additional_bytes: conf.additional_bytes,
+                fee: conf.fee,
+                lovelace: init_ada_value,
+            };
+            return Some(Self {
+                id: OnChainOrderId::from(ctx.select::<OutputRef>()),
+                pool_id: royalty_withdraw.pool_nft,
+                order: royalty_withdraw,
+            });
+        };
+        None
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct DaoRequestDataToSign {
+    pub dao_action: DaoAction,
+    pub pool_nft: PoolId,
+    pub pool_fee: u64,
+    pub treasury_fee: u64,
+    pub admin_address: Vec<InlineCredential>,
+    pub pool_address: PlutusAddress,
+    pub treasury_address: ScriptHash,
+    pub treasury_x_withdraw: i64,
+    pub treasury_y_withdraw: i64,
```

Comment:
Can it be negative?

Reply:
yep, always treasury_x_withdraw/treasury_y_withdraw <= 0. I've renamed